code/parachain/frame/pablo/src/test/dual_asset_constant_product_tests_new.rs

```diff
@@ -95,3 +95,258 @@ fn add_remove_lp() {
 		assert_eq!(Tokens::balance(lp_token, &BOB), 0_u128, "all lp tokens must have been burnt");
 	});
 }
+
+mod do_buy {
+	use composable_tests_helpers::test::helper::default_acceptable_computation_error;
+	use composable_traits::dex::{Amm, AssetAmount};
+
+	use super::*;
+
+	#[test]
+	fn should_deduct_fees_from_user_when_non_zero_fees() {
+		let initial_btc = 512_000_000_000;
+		let initial_usdt = 512_000_000_000;
+		let fee = Permill::from_rational::<u32>(3, 1000);
+		let bob_btc = 256_000;
+		let usdt_to_buy = AssetAmount::new(USDT, 128_000);
+		let fee_amount = fee.mul_ceil(usdt_to_buy.amount);
+		// 50/50 BTC/USDT Pool with a 0.3% fee
+		let init_config = PoolInitConfiguration::DualAssetConstantProduct {
+			owner: ALICE,
+			assets_weights: dual_asset_pool_weights(BTC, Permill::from_percent(50), USDT),
+			fee,
+		};
+
+		new_test_ext().execute_with(|| {
+			process_and_progress_blocks::<Pablo, Test>(1);
+
+			// Create pool
+			let pool_id = create_pool_from_config(init_config);
+
+			// Mint tokens for adding liquidity
+			assert_ok!(Tokens::mint_into(BTC, &ALICE, initial_btc));
+			assert_ok!(Tokens::mint_into(USDT, &ALICE, initial_usdt));
+
+			// Add liquidity
+			assert_ok!(Pablo::add_liquidity(
+				Origin::signed(ALICE),
+				pool_id,
+				BTreeMap::from([(BTC, initial_btc), (USDT, initial_usdt)]),
+				0,
+				false
+			));
+
+			// Mint tokens for buy
+			assert_ok!(Tokens::mint_into(BTC, &BOB, bob_btc));
+			assert_eq!(Tokens::balance(BTC, &BOB), bob_btc);
+
+			assert_ok!(Pablo::do_buy(&BOB, pool_id, BTC, usdt_to_buy, false));
+
+			// Fees deducted from buy amount in
+			assert!(default_acceptable_computation_error(
+				Tokens::balance(BTC, &BOB),
+				bob_btc - usdt_to_buy.amount - fee_amount
```

Comment:
This doesn't look right why are we subtracting a usdt amount from a btc amount ?